### ISSUE TITLE
Bokeh 0.11: Enables webgl causes inconsistent update of scatter points

### ISSUE THREAD

**Issue by siggi84** [bokeh/bokeh#3800]
This problem might be related to this issue: https://github.com/bokeh/bokeh/issues/3795

Lets consider this minimal example:
```python
from bokeh.plotting import figure, show, ColumnDataSource
from bokeh.models import Select
from bokeh.models.widgets import RadioGroup
from bokeh.io import vform
import numpy as np

N = 100
x1 = np.random.normal(0.0, 1.0,N)
x2 = x1 * (1.0 + np.random.normal(0.0, 0.1,N))
y1 = 5.0*x1 + np.random.normal(0, 0.1, N)
y2 = 50.0*x2 + np.random.normal(0, 1, N)

source = ColumnDataSource({'x' : x1, 'y' : y1})

TOOLS = "box_select,box_zoom,lasso_select,help,reset,pan"
p = figure(tools=TOOLS, plot_width=400, plot_height=400, webgl=True)
p.circle(x='x', y='y', source=source)

x_radio_group = RadioGroup(labels=["x1", "x2"], active=0)
y_radio_group = RadioGroup(labels=["y1", "y2"], active=0)
controls = vform(x_radio_group, y_radio_group)

def update(attr, old, new):
    source.data["y"] = y1 if y_radio_group.active==0 else y2
    source.data["x"] = x1 if x_radio_group.active==0 else x2
    
    # Updating the source in one step does not change this behavior
    #data = dict(x=x1 if x_radio_group.active==0 else x2, y=y1 if y_radio_group.active==0 else y2)
    #source.data = data

source.on_change('selected', update)

def select_x(value):
    update('x-change', None, None)

def select_y(value):
    update('y-change', None, None)

x_radio_group.on_click(select_x)
y_radio_group.on_click(select_y)
```

You can reproduce this problem in the following way (using the bokeh server):

- Fire up the bokeh server (e.g. bokeh server file --show)
- Select a subset of the data points (using box select or lasso)
- Switch between y1 and y2 radio buttons
- Notice that even though the y-axis scaled as expected, the scatter points did not!

Now if you repeat this exercise with webgl=False, you should see that both the y-axis and the scatter points are updated as expected.

**Comment by quebbs** [comment]
Same issue using lines.
![webgl](https://cloud.githubusercontent.com/assets/5183040/12761015/8f4dbcaa-c9a6-11e5-85f9-2a57d5b418fe.JPG)

**Comment by almarklein** [comment]
Confirmed. The problem was that the data was not updated to the GPU when there was a selection (the check whether data needs to be updated was only performed when the main glyph was rendered). Fixed in #3888.
